# Dataset Understanding

In [2]:
import pandas as pd

df = pd.read_csv("customer_churn.csv")

df.head()

,CustomerID,Age,Gender,TenureYears,MonthlyCharges,TotalCharges,ContractType,InternetService,PaymentMethod,SeniorCitizen,SupportCalls,StreamingService,PaperlessBilling,Churn
0,CUST00001,56.0,Female,0,20.41,131.92,One year,DSL,Bank transfer,No,3,No,Yes,Yes
1,CUST00002,69.0,Male,9,21.10,2394.45,Two year,DSL,Credit card,Yes,1,Yes,No,No
2,CUST00003,46.0,Female,3,61.66,2755.45,Two year,DSL,Bank transfer,No,3,Yes,No,No
3,CUST00004,32.0,Male,8,35.50,3796.58,Month-to-month,No,Credit card,No,2,No,No,No
4,CUST00005,60.0,Male,9,86.45,10188.80,Two year,Fiber optic,Electronic check,Yes,3,No,Yes,Yes


#### Dataset Name

Customer Churn Dataset

#### Dataset Source

Sprint 4 practice dataset created for this project from a simulated real-world telecom customer scenario.

| Column             | Type                 | Meaning                              |
| ------------------ | -------------------- | ------------------------------------ |
| `CustomerID`       | Identifier           | Unique customer ID                   |
| `Age`              | Numerical            | Customer's age                       |
| `Gender`           | Categorical          | Customer gender                      |
| `TenureYears`      | Numerical/Discrete   | Number of years customer has stayed  |
| `MonthlyCharges`   | Numerical/Continuous | Monthly amount paid                  |
| `TotalCharges`     | Numerical/Continuous | Total amount paid                    |
| `ContractType`     | Categorical          | Customer's contract type             |
| `InternetService`  | Categorical          | Internet service used                |
| `PaymentMethod`    | Categorical          | Payment method                       |
| `SeniorCitizen`    | Categorical          | Whether customer is a senior citizen |
| `SupportCalls`     | Numerical/Discrete   | Number of support calls              |
| `StreamingService` | Categorical          | Whether customer uses streaming      |
| `PaperlessBilling` | Categorical          | Whether paperless billing is enabled |
| `Churn`            | Categorical/Target   | Whether customer left the company    |


#### Target variable: Churn

* Yes → Customer left
* No → Customer stayed

### 1.Dataset

A dataset is a structured collection of data containing multiple observations or records and their corresponding characteristics.

Real-world example:

A telecom company may maintain customer information such as:

* Age
* Contract type
* Monthly charges
* Internet service
* Payment method
* Customer churn

Each customer represents one record, while each characteristic represents a feature.

In [5]:
df.head(10)

,CustomerID,Age,Gender,TenureYears,MonthlyCharges,TotalCharges,ContractType,InternetService,PaymentMethod,SeniorCitizen,SupportCalls,StreamingService,PaperlessBilling,Churn
0,CUST00001,56.0,Female,0,20.41,131.92,One year,DSL,Bank transfer,No,3,No,Yes,Yes
1,CUST00002,69.0,Male,9,21.10,2394.45,Two year,DSL,Credit card,Yes,1,Yes,No,No
2,CUST00003,46.0,Female,3,61.66,2755.45,Two year,DSL,Bank transfer,No,3,Yes,No,No
3,CUST00004,32.0,Male,8,35.50,3796.58,Month-to-month,No,Credit card,No,2,No,No,No
4,CUST00005,60.0,Male,9,86.45,10188.80,Two year,Fiber optic,Electronic check,Yes,3,No,Yes,Yes
5,CUST00006,25.0,Male,5,90.88,5818.04,Two year,DSL,Mailed check,No,0,No,No,No
6,CUST00007,38.0,Female,5,37.79,2670.55,Month-to-month,DSL,Mailed check,No,1,Yes,Yes,No
7,CUST00008,56.0,Male,0,108.85,1106.71,Two year,No,Electronic check,No,4,Yes,No,Yes
8,CUST00009,36.0,Male,9,64.19,7652.76,One year,DSL,Mailed check,No,3,Yes,No,No
9,CUST00010,40.0,Male,1,34.28,618.59,One year,Fiber optic,Electronic check,No,3,Yes,Yes,No


### 2.Feature

A feature is an individual variable or column that describes some characteristic of an observation.

For example:

* Age
* MonthlyCharges
* ContractType
* InternetService
* PaymentMethod

These are all features of a customer.

Real-world example:

For predicting whether a customer will leave a telecom company:

* Age → feature
* Monthly charges → feature
* Contract type → feature
* Internet service → feature

In [6]:
features = df.columns

print(features)

Index(['CustomerID', 'Age', 'Gender', 'TenureYears', 'MonthlyCharges',
       'TotalCharges', 'ContractType', 'InternetService', 'PaymentMethod',
       'SeniorCitizen', 'SupportCalls', 'StreamingService', 'PaperlessBilling',
       'Churn'],
      dtype='str')


### 3.Target

The target is the variable that we want to predict or explain.

In our dataset:

Churn is the target.

Real-world example:

A telecom company wants to predict:

"Will this customer leave the company?"

Therefore:

Target = Churn

In [7]:
target = "Churn"

print(df[target].head())

0    Yes
1     No
2     No
3     No
4    Yes
Name: Churn, dtype: str


### 4. Independent Variables

Independent variables are variables used as inputs to predict the target.

They are also called:

* Input variables
* Predictors
* Features

For our dataset:

* Age
* Gender
* TenureYears
* MonthlyCharges
* TotalCharges
* ContractType
* InternetService
* PaymentMethod

are independent variables.

Real-world example:

We could use: Age + ContractType + MonthlyCharges + TenureYears

to predict: Churn

In [8]:
X = df.drop("Churn", axis=1)

print(X.head())

  CustomerID   Age  Gender  TenureYears  MonthlyCharges  TotalCharges  \
0  CUST00001  56.0  Female            0           20.41        131.92   
1  CUST00002  69.0    Male            9           21.10       2394.45   
2  CUST00003  46.0  Female            3           61.66       2755.45   
3  CUST00004  32.0    Male            8           35.50       3796.58   
4  CUST00005  60.0    Male            9           86.45      10188.80   

     ContractType InternetService     PaymentMethod SeniorCitizen  \
0        One year             DSL     Bank transfer            No   
1        Two year             DSL       Credit card           Yes   
2        Two year             DSL     Bank transfer            No   
3  Month-to-month              No       Credit card            No   
4        Two year     Fiber optic  Electronic check           Yes   

   SupportCalls StreamingService PaperlessBilling  
0             3               No              Yes  
1             1              Yes          

### 5. Dependent Variable

A dependent variable is the variable whose value depends on the independent variables.

In machine learning, this is generally the target variable.

For our dataset:

Dependent Variable = Churn

In [10]:
y = df["Churn"]

print(y.head())

0    Yes
1     No
2     No
3     No
4    Yes
Name: Churn, dtype: str


### 6. Numerical Features

Numerical features contain numbers and can be used for mathematical calculations.

Examples from our dataset:

* Age
* TenureYears
* MonthlyCharges
* TotalCharges
* SupportCalls

Real-world example:

MonthlyCharges could contain:

45.50,
72.30,
99.99

We can calculate:

* Average
* Minimum
* Maximum
* Standard deviation

In [11]:
numerical_columns = df.select_dtypes(include="number").columns

print(numerical_columns)

Index(['Age', 'TenureYears', 'MonthlyCharges', 'TotalCharges', 'SupportCalls'], dtype='str')


In [12]:
df[numerical_columns].describe()

,Age,TenureYears,MonthlyCharges,TotalCharges,SupportCalls
count,4969.000000,5020.000000,4970.000000,4969.000000,5020.000000
mean,43.580600,5.043426,70.085469,4679.134800,1.988048
std,14.922491,3.162546,28.852719,3468.389578,1.416837
min,18.000000,0.000000,20.020000,29.350000,0.000000
25%,31.000000,2.000000,45.920000,1883.140000,1.000000
50%,43.000000,5.000000,70.165000,3830.980000,2.000000
75%,56.000000,8.000000,94.957500,6917.120000,3.000000
max,69.000000,10.000000,119.980000,15559.290000,9.000000


### 7. Categorical Features

Categorical features contain categories or labels instead of continuous numerical measurements.

Examples:

* Gender
* ContractType
* InternetService
* PaymentMethod
* StreamingService
* PaperlessBilling

Real-world example:

ContractType can have:

* Month-to-month
* One year
* Two year

These values represent categories.

In [14]:
categorical_columns = df.select_dtypes(include="str").columns

print(categorical_columns)

Index(['CustomerID', 'Gender', 'ContractType', 'InternetService',
       'PaymentMethod', 'SeniorCitizen', 'StreamingService',
       'PaperlessBilling', 'Churn'],
      dtype='str')


### 8. Discrete Features

Discrete features contain countable values, usually whole numbers.

Examples:

* Number of support calls
* Number of purchases
* Number of children
* Number of transactions

In our dataset:

SupportCalls is a discrete numerical feature.

Example:
0

1

2

3

4

5

You cannot normally have:

2.73 support calls

In [15]:
print(df["SupportCalls"].head(10))

0    3
1    1
2    3
3    2
4    3
5    0
6    1
7    4
8    3
9    3
Name: SupportCalls, dtype: int64


### 9. Continuous Features

Continuous features can take a range of numerical values, including decimal values.

Examples:

* Height
* Weight
* Temperature
* Salary
* MonthlyCharges

In our dataset:

* MonthlyCharges
* TotalCharges

are continuous features.

Example:

49.99

72.45

89.75

103.20

In [16]:
print(df["MonthlyCharges"].head(10))

0     20.41
1     21.10
2     61.66
3     35.50
4     86.45
5     90.88
6     37.79
7    108.85
8     64.19
9     34.28
Name: MonthlyCharges, dtype: float64


### 10. Identifier Columns

An identifier column uniquely identifies each record.

In our dataset:

CustomerID is an identifier.

Example:

CUST00001

CUST00002

CUST00003

#### Important ML concept:

An identifier generally should not be used as a predictive feature because it doesn't represent meaningful customer behavior.

For example:

CustomerID = CUST00001

doesn't tell us whether the customer will churn.

In [17]:
print(df["CustomerID"].head())

0    CUST00001
1    CUST00002
2    CUST00003
3    CUST00004
4    CUST00005
Name: CustomerID, dtype: str


### 11. Target Variable

The target variable is the output that a machine learning model tries to predict.

Our target:

Churn contains:

* Yes
* No

This is a classification problem because the target contains categories.

Real-world example:

A telecom company could use the model to identify customers who are likely to leave.

Customer information

        ↓
        
Machine Learning Model

        ↓
        
Churn prediction

        ↓
        
Yes / No

In [18]:
df["Churn"].value_counts()

Churn
No     3623
Yes    1397
Name: count, dtype: int64

In [19]:
df["Churn"].value_counts(normalize=True) * 100

Churn
No     72.171315
Yes    27.828685
Name: proportion, dtype: float64

### 12. Metadata

Metadata means information that describes the dataset and its columns.

Examples:

* Dataset name
* Number of rows
* Number of columns
* Column names
* Data types
* Missing values
* Description of columns
* Target variable
* Data source

In [20]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5020 entries, 0 to 5019
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerID        5020 non-null   str    
 1   Age               4969 non-null   float64
 2   Gender            5020 non-null   str    
 3   TenureYears       5020 non-null   int64  
 4   MonthlyCharges    4970 non-null   float64
 5   TotalCharges      4969 non-null   float64
 6   ContractType      5020 non-null   str    
 7   InternetService   5020 non-null   str    
 8   PaymentMethod     5020 non-null   str    
 9   SeniorCitizen     5020 non-null   str    
 10  SupportCalls      5020 non-null   int64  
 11  StreamingService  5020 non-null   str    
 12  PaperlessBilling  5020 non-null   str    
 13  Churn             5020 non-null   str    
dtypes: float64(3), int64(2), str(9)
memory usage: 814.5 KB


### 13. Dataset Dimensions

Dataset dimensions describe the number of rows and columns.

For example:

Rows = 5020

Columns = 14

In [21]:
df.shape

(5020, 14)

In [22]:
rows, columns = df.shape

print("Number of rows:", rows)
print("Number of columns:", columns)

Number of rows: 5020
Number of columns: 14
